### ISSUE TITLE
Release CUDA binary (v0.48.0) never compiles FlashAttention-2; fused kernels also skipped on f16 fine-tunes

### ISSUE THREAD

**Issue by vijaydevmasters** [f-inverse/jammi-ai#443]
Release CUDA binary (v0.48.0) never compiles FlashAttention-2; fused kernels also skipped on f16 fine-tunes

## Summary

The published `jammi-server-cu12` 0.48.0 binary never runs FlashAttention-2. `FLASH_COMPILED` is `false` because the CUDA release lane builds `--features cuda,jetstream-broker,storage-cloud` and **does not** enable `flash-attn`. That is compile-time, not a runtime miss.

Separately, even Jammi's own fused kernels (`attention_block_fused`, `layer_norm_fused`, `rope_fused`, `geglu_fused`, `lora_linear_fused`, `softmax_last_dim_fused`) decline on `backbone_dtype="f16"` and fall back to eager composition. FA2 is also **bf16-only** at admission, so an f16 fine-tune would still miss FA after a flash-compiled rebuild.

Observed on a real A100 80GB fine-tune of BioClinical-ModernBERT-large (seq_len=128, batch=16) against `jammi-server 0.48.0`.

## What the release actually builds

`.github/workflows/release-binaries.yml` (`server-cu12` job):

```
cargo build --release -p jammi-server --bin jammi-server --features cuda,jetstream-broker,storage-cloud
```

Same feature list in `.github/workflows/pypi-server-cuda.yml` (`cargo_features: cuda,jetstream-broker,storage-cloud`).

Checkout is plain `actions/checkout@v4` — CUTLASS is not initialized. `crates/jammi-kernels/build.rs::build_flash_attn` panics if `third_party/cutlass/include/cutlass/cutlass.h` is missing, so this lane cannot compile FA2 even if the feature were added without also checking out the submodule.

This is intentional today:

- `crates/jammi-kernels/Cargo.toml`: `flash-attn = ["cuda"]`, and `cuda` must **not** imply `flash-attn` (CUTLASS + nvcc).
- `ci/scripts/check_flash_attn_closure.py` **fails CI** if `jammi-server --features cuda,jetstream-broker,storage-cloud` ever reaches `jammi-kernels/flash-attn`.

So v0.48.0 vendors FA2 in-tree, but the shipping CUDA tarball/`jammi-server-cu12` wheel is `FLASH_COMPILED=false`.

## Runtime path (why nothing "turns on")

`crates/jammi-kernels/src/admission.rs`:

```
pub const FLASH_COMPILED: bool = cfg!(feature = "flash-attn");
```

`crates/jammi-encoders/src/modernbert.rs` (`flash_capability_gates`): first check is `feature_compiled`. On this binary it returns `CapabilityMiss` / `"flash_attn_feature_compiled"` and never reaches the FA2 kernel.

Cascade is:

1. `attention_block_flash` — compiled out (`FLASH_COMPILED=false`)
2. `attention_block_fused` — domain miss on f16 (`dtype_f32_or_bf16_*`)
3. eager composition — what actually ran

Logs from the 0.48.0 server on first forward (f16 LoRA fine-tune):

```
fused-kernel domain check failed; falling back to the eager composition op="layer_norm_fused" predicate="dtype_f32_or_bf16_matching_between_x_and_weight"
fused-kernel domain check failed; falling back to the eager composition op="lora_linear_fused" predicate="base_dtype_f32_or_bf16_matched"
fused-kernel domain check failed; falling back to the eager composition op="attention_block_fused" predicate="dtype_f32_or_bf16_matching_between_qkv_and_mask"
fused-kernel domain check failed; falling back to the eager composition op="rope_fused" predicate="dtype_f32_or_bf16_matching_between_x_cos_sin"
fused-kernel domain check failed; falling back to the eager composition op="softmax_last_dim_fused" predicate="dtype_f32_or_bf16_matching_between_scores_and_mask"
fused-kernel domain check failed; falling back to the eager composition op="geglu_fused" predicate="dtype_f32_or_bf16"
```

No `attention_block_flash` line appears because flash admission never Holds.

`pip install flash-attn` does not help: that is the PyTorch wheel. This server is Candle/Rust and does not load it.

## Code that has to change for FA2 to trigger in the shipping binary

Do **not** fold `flash-attn` into the `cuda` feature. Keep it opt-in and add it to the **release cargo command**. Folding it into `cuda` is exactly the leak `check_flash_attn_closure.py` exists to prevent.

| File | Change |
|---|---|
| `crates/jammi-ai/Cargo.toml` | Add `flash-attn = ["cuda", "jammi-encoders/flash-attn", "jammi-kernels/flash-attn"]`. `cuda` currently stops at `jammi-kernels/cuda` / `jammi-encoders/cuda` and does not forward FA. |
| `crates/jammi-server/Cargo.toml` | Add `flash-attn = ["jammi-ai/flash-attn"]`. Today only `cuda = ["jammi-ai/cuda"]` exists. |
| `crates/jammi-encoders/Cargo.toml` | Already has `flash-attn = ["jammi-kernels/flash-attn"]`. No change. |
| `.github/workflows/release-binaries.yml` | Build with `--features cuda,flash-attn,jetstream-broker,storage-cloud`. Checkout CUTLASS (`submodules: true` or `git submodule update --init --depth 1 crates/jammi-kernels/third_party/cutlass`). |
| `.github/workflows/pypi-server-cuda.yml` | Same feature list. |
| `ci/scripts/check_flash_attn_closure.py` | Stop treating that CUDA **release** lane as CUTLASS-forbidden, or CI will fail the PR. Keep the guard that `cuda` alone must not imply `flash-attn`. |

`build.rs` already has the rest (nvcc, CUTLASS path, gencode for sm80/86/89/90). A100 sm80 is in the validated set.

Local equivalent:

```
cargo build --release -p jammi-server --bin jammi-server \
  --features cuda,flash-attn,jetstream-broker,storage-cloud
```

## Runtime gate that still blocks FA after a flash-compiled rebuild

`flash_capability_gates` (same file) then requires:

- CUDA device
- arch in the FA validated set (sm80/86/89/90 — A100 OK)
- **`dtype == BF16` only** (`"dtype_is_bf16"`)
- head dim 64

An f16 fine-tune (`backbone_dtype="f16"`) would still decline FA with `dtype_is_bf16`, and would still miss fused kernels (`dtype_f32_or_bf16_*`).

If the goal is "FA2 actually runs on the default CUDA server for encoder fine-tunes", either:

- document that FA/fused need `backbone_dtype="bf16"`, and/or
- extend flash + fused admission to f16 if that is a supported training dtype.

Short-seq encoder training (b16·s128) may not need FA2 for speed (measured FA2 lift was small at b8·s512). The fused kernels are the larger win and already ship in this tarball — they are skipped only because of f16.

**Comment by thankrandomness** [comment]
## Campaign handoff — operator-approved, for a fresh session to drive to close

**Operator decision (Vijay, 2026-08-31):** this campaign is approved and scoped as follows. (1) Ship FA2 in the release artifacts. (2) Build the release-artifact capability gate (the class fix). (3) **f16 acceleration is IN scope: validate and build it** — not merely document bf16-first. The operator explicitly overrode the document-only option: "jammi should absolutely validate f16 acceleration and build it if it's not already accelerating it." (4) Submission-time dtype visibility regardless.

Everything below was verified in-tree at `main` = 309d1a10 (plus `gh` reads) by a dedicated review; citations are to that tree. Re-verify citations before acting — the #435 campaign branch (`feat/351-gpu-validation`, closes #433/#434) may have merged by the time you read this; **base this campaign on main after #435 lands.**

### Verified facts (do not re-derive; do re-cite)

**The issue body above checks out on every point.** Additional verified context:

1. **What the perf marathon measured vs what ships.** The fine-tune/attention perf producers build jammi-bench WITH flash: `cargo build --release -p jammi-bench --features cuda,jammi-encoders/flash-attn` (`ci/scripts/perf/finetune_ab.sh:616`, `finetune_run_ab.sh:305`, `stacked_sweep.sh:324`, `fa2_ab.sh:7`, `clip_artifact_producer.sh:80`). Committed goldens bake the proof: `ci/scripts/perf/fixtures/finetune_run_golden/modernbert_fused.json:12-15` records `"build_features": ["bench-cuda","cuda","flash-attn"]` (baked by `crates/jammi-bench/src/report.rs:199-210` from `jammi_kernels::admission::{CUDA_COMPILED,FLASH_COMPILED}`). Measurements ran `JAMMI_KERNELS_STRICT=1` (`admission.rs:510-517`) so they are genuinely fused/flash-path numbers. Dtypes: bf16 (training/attention), f32 (CPU/inference); **no perf lane ever ran f16**. Server default `compute_precision` is F32 (`crates/jammi-db/src/config.rs:396-402`).
2. **The closure-gate mechanism that created the gap.** `check_flash_attn_closure.py:63-64` bans the release feature set from reaching flash-attn; `:89-91` carries `ALL_FEATURES_FLASH_EXEMPT = {"jammi-encoders": ["jammi-kernels/flash-attn"]}` — the declared exemption for the measurement lanes. The release lane (`release-binaries.yml:216`, `pypi-server-cuda.yml:46`) builds `--features cuda,jetstream-broker,storage-cloud` with plain `actions/checkout@v4` (no CUTLASS submodule). `crates/jammi-kernels/build.rs::build_flash_attn` panics without `third_party/cutlass/include/cutlass/cutlass.h`, so the feature flag alone is insufficient — the lane needs the submodule too. FA2 is not crates.io-publishable (`jammi-kernels/Cargo.toml:29-31`).
3. **No CI lane runs the artifact users get — confirmed.** The release-blocking A100 gate (`_gpu-prove-gate.yml`, required by `release-binaries.yml:181-201`) builds its own test binaries with different feature sets (`ci/scripts/runpod_gpu_prove.sh:91-106`: `cuda,live-gpu-tests`, no broker/storage triple). The `server-cu12` build job runs on a GPU-less runner ("nothing runs the binary here", `release-binaries.yml:180`). The tag-blocking perf gate (`crates.yml:67-112`) is CPU-hermetic. The only shipped-artifact guards are static (`check_flash_attn_closure.py`) or run on the user's machine (the load-time quantized-cuda canary).
4. **What ships under plain `cuda`:** all six fused kernels (`AttentionBlockFused`, `LayerNormFused`, `RopeFused`, `GegluFused`, `LowRankResidualLinear`, `SoftmaxLastDimFused`), plus `MemEfficientAttention`, `RopePositionsFused`, `adamw_step_fused`, `DropoutFused`, cast ops, `QuantMatMulGrad`/GGUF/QLoRA (`crates/jammi-kernels/src/ops/mod.rs:141-208`; only the flash module is `#[cfg(feature = "flash-attn")]`, lines 98/184-193). **Compiled out:** the whole FA2 program — the #389 dense arm (1.071× torch-sdpa @ bf16), #408 varlen-RoPE, #411 padded→flash dispatch.
5. **Admission predicates (the f16 wall):** every fused predicate admits `F32 | BF16` only (`crates/jammi-encoders/src/layer_norm.rs:118`, `modernbert.rs:263,636,738,818,1842`); FA2 additionally requires exactly BF16 (`modernbert.rs` `flash_capability_gates` → `DomainMiss "dtype_is_bf16"`; FFI domain check `jammi-kernels/src/ops/flash_attention.rs:102,116`) and head-dim 64. First gate of the cascade is `FLASH_COMPILED = cfg!(feature="flash-attn")` (`admission.rs:1098`).
6. **Hardware note:** on the validated arch set (sm_80/86/89/90) tensor cores run f16 and bf16 at identical throughput; f16's exclusion is a software domain-of-validity choice, not a hardware limit. Upstream FlashAttention-2 supports fp16 natively (it was its original dtype) — jammi's admission is what narrows to bf16. `build.rs` pins the 4-arch `GENCODE_ARCHES` fat set and neutralizes `CUDA_COMPUTE_CAP` (`build.rs:365-380`), so the container lane's `CUDA_COMPUTE_CAP=80` base env does not narrow SASS.

### The fix plan (four parts)

**Part 1 — ship FA2** (mechanical; the issue body's table is correct, keep it):
- `crates/jammi-ai/Cargo.toml`: `flash-attn = ["cuda", "jammi-encoders/flash-attn", "jammi-kernels/flash-attn"]`.
- `crates/jammi-server/Cargo.toml`: `flash-attn = ["jammi-ai/flash-attn"]`.
- Both release lanes: `--features cuda,flash-attn,jetstream-broker,storage-cloud` + CUTLASS submodule checkout.
- `check_flash_attn_closure.py`: re-scope so the release lane is flash-permitted while KEEPING the invariant that `cuda` alone never implies `flash-attn` (that ban is the gate's real job; do not weaken it). **Do NOT fold `flash-attn` into `cuda`.**
- Container lane (`Dockerfile` runtime-cuda variant / `server-image.yml`): same treatment — the image must ship the same capability surface as the wheel/tarball.

**Part 2 — release-artifact capability gate** (the class fix; without it Part 1 regresses silently):
A required lane that builds with the EXACT release feature set (ideally executes the actual packaged artifact) on a GPU pod and asserts the capability surface: `FLASH_COMPILED == true`, flash admission Holds on a bf16 (and, after Part 3, f16) probe, all six fused admissions Hold. This is the artifact-axis analog of `ci/scripts/check_gpu_parity_matrix.py` (the silicon-axis completeness gate born from the #434 class): proof surface must equal shipped surface on the FEATURE axis too. Acceptance: the gate is RED against today's release-lane config and GREEN after Part 1.

**Part 3 — f16 acceleration, validated and built** (operator-directed; the big half):
- Extend the six fused kernels' admission predicates and FA2 admission to F16, each extension carrying the same oracle rigor every admitted dtype has: CPU-reference parity oracles at f16, overflow/underflow-boundary oracles (f16 normal range ≈ 6.1e-5 … 65504 — test AT the boundary, not just inside it), gradient-finiteness oracles, and the KO-1..KO-8 kernel-acceptance standard (see `#393` / the KO registry `ci/kernel-oracle-helpers.txt` mechanism).
- **Design fork to pressure-test before any code:** does f16 *training* need a loss-scaling apparatus (GradScaler equivalent: f32 master weights + dynamic scale + overflow-step skipping), or do bounded encoder fine-tunes train stably in raw f16? PyTorch's answer is "AMP + GradScaler"; jammi has no such machinery today. The pressure-tester should demand measured evidence either way (run the fixture fine-tunes at f16, watch for inf/NaN) before the trainer design is fixed. Do not extend admission without settling this — an f16 path that NaNs on real models is worse than the current honest fallback.
- FA2 f16: the upstream kernel supports fp16; validation burden is jammi-side admission + parity + the head-dim-64 domain on all four arches.
- Perf A/B: extend the bench producers (`finetune_ab.sh` etc.) with f16 legs so f16's gains are MEASURED like every other claim (the perf-unification provenance machinery — build_sha stamping, claim-binding gate — applies).
- 4-arch validation is mandatory: sm_80/86/89/90 via the existing prove machinery (`ci/scripts/runpod_gpu_prove.sh`, `GPU_PROVE_ARCH`, per-leg `NATIVE_COMPUTE_CAP`, `l4_l40s` capacity fallback, requested-vs-provisioned assert exit 97).

**Part 4 — submission-time dtype visibility** (ship regardless of Part 3's timeline): when a fine-tune's dtype excludes fused/flash paths, say so AT SUBMISSION with the remedy, not one log line per op after the first forward.

### House constraints the fresh session must honor

- **Single consolidated PR** for the whole campaign (operator standing rule). Full rigor chain: gap-analyzer → issue-triage (this issue is the raw input; classify + symptom_spec → escape row) → pressure-tester on the design (especially the Part-3 loss-scaling fork) → owner waves → adversarial-audit rounds → fix-verifier RED→GREEN (Part 2's gate red-on-main IS the red) → oracle → acceptance-verifier. Lead-gate relay discipline on any BLOCK.
- **GPU execution is mandatory, not optional** — baselines on real silicon, all four arches, before close.
- **Proof surface == shipped surface** (correct-over-preexisting): the entire point of this campaign; also why Part 2 is non-negotiable.
- **Extend seams, not upstream**: no upstream issues (candle/FA2); wrap at house seams; track upstream state as ledger context only.
- **Fold-the-fix**: anything found en route gets fixed in the campaign PR, not filed-and-escaped.
- **Engine↔cookbook co-evolution**: a chapter unit exercising FA2 + f16 fine-tuning through the public Python surface (22-precision style: live cells, measured pinned numbers, honest limits) ships in the same campaign; chapter-authoring findings route back as defects.
- **Wave hygiene**: per-wave `CARGO_TARGET_DIR` under `$HOME/.cache/jammi-ct/`, deleted at wave end, confirmed in report; lead sweeps between waves. Commit before mutating anything in a worktree.
- `SWARM_GATE_TOUCHED` human-amend-only gates (`check_flash_attn_closure.py` likely qualifies — check): the operator's PR merge is the human amendment; state it in the PR body.
- Coordinate with PR #435 (`feat/351-gpu-validation`): it closes #433/#434 and reworks admission/canary/KO-7 surfaces this campaign touches. Base on main AFTER it merges; if it hasn't merged yet, stop and ask the operator rather than forking the surface.

### Acceptance criteria (the campaign is done when)

1. The Part-2 gate exists, is required, was RED against the pre-fix release config, and is GREEN at the tip — asserting `FLASH_COMPILED`, flash admission Holds (bf16 + f16), and all six fused admissions Hold on the release-feature-set binary on a real GPU.
2. An f16 fine-tune of the standard fixture engages all six fused kernels and FA2 (admission logs show Holds, `JAMMI_KERNELS_STRICT=1` run passes), with f16 parity/boundary/finiteness oracles green on all four arches.
3. f16 perf legs measured and committed through the provenance machinery; the issue's reporter scenario (BioClinical-ModernBERT-large, A100, b16·s128) re-run showing fused+flash engagement.
4. The submission-time dtype notice ships with a test pinning its wording/trigger.
5. Cookbook chapter live-renders the f16/FA2 path with measured numbers.

*Handoff prepared by the #435 session lead; review evidence gathered read-only from main @ 309d1a10. The reporter's analysis in the issue body was independently verified — trust it.*
